File: src/main/IDX.cpp
#include "IDX.hpp"

using namespace std;

void getIdxHeaders(int fdInputFile, int& magicNumber, int& numImages, int& numRows, int& numColumns) {
    // Read magic number
    if (read(fdInputFile, &magicNumber, sizeof(int)) < ssize_t(sizeof(int)))
        errExit("read");
    magicNumber = be32toh(magicNumber);
    // cout << "magic number = " << magicNumber << endl;

    // Read number of images
    if (read(fdInputFile, &numImages, sizeof(int)) < ssize_t(sizeof(int)))
        errExit("read");
    numImages = be32toh(numImages);
    // cout << "number of images = " << numImages << endl;

    // Read number of rows
    if (read(fdInputFile, &numRows, sizeof(int)) < ssize_t(sizeof(int)))
        errExit("read");
    numRows = be32toh(numRows);
    // cout << "number of rows = " << numRows << endl;

    // Read number of columns
    if (read(fdInputFile, &numColumns, sizeof(int)) < ssize_t(sizeof(int)))
        errExit("read");
    numColumns = be32toh(numColumns);
    // cout << "number of columns = " << numColumns << endl;
}




void getIdxLabelHeaders(int fdInputFile,
    int& magicNumber, 
    int& numLabels) {

    // Read magic number
    if (read(fdInputFile, &magicNumber, sizeof(int)) < ssize_t(sizeof(int)))
        errExit("read");
    magicNumber = be32toh(magicNumber);
    cout << "magic number = " << magicNumber << endl;

    // Read number of images
    if (read(fdInputFile, &numLabels, sizeof(int)) < ssize_t(sizeof(int)))
        errExit("read");
    numLabels = be32toh(numLabels);
    cout << "number of labels = " << numLabels << endl;

}




std::vector<Image> getIdxData(const char *filename, int pixel_size, int read_upto, int* img_size) {
    // Open file
    int fd;
    if ((fd = open(filename, O_RDONLY)) == -1)
        errExit("IDX.getIdxData.open(filename)");

    // Read IDX headers
    int magicNumber;
    int numImages;
    int numRows;
    int numColumns;
    getIdxHeaders(fd, magicNumber, numImages, numRows, numColumns);

    Image img(numRows, numColumns, pixel_size);
    std::vector<Image> ret;

    // Read images
    for (int i=0; i<read_upto; i++) {
        img.scan(fd, i+1);
        ret.push_back(img);
    }

    close(fd);

    *img_size = numRows*numColumns;
    return ret;
}



std::vector<int> getIdxLabelData(const char *filename,  
    int read_upto) {

    // Open file
    int fd;
    if ((fd = open(filename, O_RDONLY)) == -1)
        errExit("IDX.getIdxLabelData.open(filename)");

    // Read IDX headers
    int magicNumber;
    int numLabels;
    getIdxLabelHeaders(fd, magicNumber, numLabels);

    std::vector<int> ret(numLabels, 0);

    // Read labels
    for (int i=0; i<read_upto; i++) 
        if (read(fd, &ret[i], sizeof(char)) < ssize_t(sizeof(char)))
            errExit("IDX.getIdxLabelData.read:");

    close(fd);
    return ret;
}

File: python/idx.py
import unittest
from Image import Image



def getIdxHeaders(f):
   magicnumber = f.read(4)
   magicnumber = int.from_bytes(magicnumber, "big")

   numofimages = f.read(4)
   numofimages = int.from_bytes(numofimages, "big")

   numrows = f.read(4)
   numrows = int.from_bytes(numrows, "big")

   numcols = f.read(4)
   numcols = int.from_bytes(numcols, "big")

   return (magicnumber, numofimages, numrows, numcols)




def getIdxLabelHeaders(f):

    magicnumber = f.read(4)
    magicnumber = int.from_bytes(magicnumber, "big")

    numlabels = f.read(4)
    numlabels = int.from_bytes(numlabels, "big")

    return (magicnumber, numlabels)




# Assumed that headers have been already read from file f
def getIdxData(f, num_elements, num_rows, num_cols, element_size):
    
    ret = [[[0] * num_cols for i in range(num_rows)] for i in range(num_elements)]

    for e in range(num_elements):
        for i in range(num_rows):
            for j in range(num_cols):
                ret[e][i][j] = f.read(element_size)
                ret[e][i][j] = int.from_bytes(ret[e][i][j], "big")

    return ret


    

# Assumed that headers have been already read from file f
def getIdxLabels(f, num_labels, label_size):

    ret = [0 for i in range(num_labels)]
    
    for i in range(num_labels):
        ret[i] = f.read(label_size)
        ret[i] = int.from_bytes(ret[i], "big")

    return ret
    



def print_upto(filename, upto):
    with open(filename, "rb") as f:
        magicnumber, numofimages, numrows, numcols = getIdxHeaders(f)
        if upto > numofimages:
            upto = numofimages
        img = Image(numrows, numcols)
        for i in range(upto):
            img.scan(f, id=i+1, normalize=False)
            img.print()




def print_img(filename, img_num):
    with open(filename, "rb") as f:
        magicnumber, numofimages, numrows, numcols = getIdxHeaders(f)
        img = Image(numrows, numcols)
        for i in range(numofimages):
            img.scan(f, id=i+1, normalize=False)
            if i == img_num-1:
                img.print() 
                return




class Test(unittest.TestCase):

    @classmethod
    def setUpClass(self):

        self.tests_folder = './test_input/'
        self.data_file_1 = self.tests_folder + 't10k-images-idx3-ubyte'
        self.label_file_1 = self.tests_folder + 't10k-labels-idx1-ubyte'


    # Just prints, does not actually test
    def test_1(self):

        read_upto = 10

        f = open(self.data_file_1, "rb")
        magicnumber, numofimages, numrows, numcols = getIdxHeaders(f)
        ret = getIdxData(f, read_upto, numrows, numcols, 1)
        f.close()
        
        images = [Image(numrows, numcols, i+1) for i in range(read_upto)]
        for i in range(read_upto):
            images[i].pixels = ret[i]
        

        f = open(self.label_file_1, "rb")
        magicnumber, numlabels = getIdxLabelHeaders(f)
        ret = getIdxLabels(f, read_upto, 1)
        f.close()

        labels = [0 for i in range(read_upto)]
        for i in range(read_upto):
            labels[i] = ret[i]

        # Print
        for i in range(read_upto):
            print("Label:", labels[i])
            images[i].print()




if __name__ == "__main__":
    unittest.main()